### PR TITLE
Refactor cluster cache.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**
@siddontang  PTAL

### CODE REVIEW

File: server/cache.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+	store, _ := c.stores[storeID]
+	return store
+}
+
+func (c *ClusterInfo) getStores() map[uint64]*StoreInfo {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.stores
```
Comment (siddontang) [inline server/cache.go]: seems that we may still meet concurrent problem when we use stores outer. 
Lock does nothing.

Comment (qiuyesuifeng) [inline server/cache.go]: maybe needless, except for tests, if u call getStores to get  stores, that will always be only readable.

Comment (siddontang) [inline server/cache.go]: get stores is readable, but add/remove store can change it.

File: server/cache.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			}
+		}
+	}
+
+	c.regions[regionID] = &RegionInfo{
+		region: region,
+		peer:   leaderPeer,
+	}
+
+	if store, ok := c.stores[leaderPeer.GetStoreId()]; ok {
```
Comment (siddontang) [inline server/cache.go]: if not ok, no update store regions?

File: server/cache.go (hunk 1 above)
Comment (disksing) [inline server/cache.go]: I also think it's not safe. Better to dump them into a `[]*StoreInfo`.

File: server/cache.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+	delete(c.stores, storeID)
+}
+
+func (c *ClusterInfo) getStore(storeID uint64) *StoreInfo {
+	c.RLock()
+	defer c.RUnlock()
+
+	store, _ := c.stores[storeID]
+	return store
```
Comment (disksing) [inline server/cache.go]: `return c.stores[storeID]` should be enough.

File: server/cache.go (hunk 2 above)
Comment (qiuyesuifeng) [inline server/cache.go]: It is invalid if there is not store here, so if the store is not existed, we may do nothing.

**Comment by qiuyesuifeng** [comment]
PTAL

File: server/cache.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import (
+	"math/rand"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
+type storeIDMap map[uint64]struct{}
```
Comment (siddontang) [inline server/cache.go]: seem no used.

Comment (qiuyesuifeng) [inline server/cache.go]: yes, it can be removed.

File: server/cache.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+	for regionID := range storeRegions {
+		if idx == randIdx {
+			randRegionID = regionID
+			break
+		}
+
+		idx++
+	}
+
+	return r.leaderRegions[randRegionID]
```
Comment (disksing) [inline server/cache.go]: Need `clone()` for `RegionInfo`?

Comment (qiuyesuifeng) [inline server/cache.go]: Yes, that is more safe.

File: server/cache.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+func (s *StoreInfo) clone() *StoreInfo {
+	return &StoreInfo{
+		store: proto.Clone(s.store).(*metapb.Store),
+		stats: proto.Clone(s.stats).(*pdpb.StoreStats),
+	}
+}
+
+// usedFraction is the used fraction of storage capacity.
+func (s *StoreInfo) usedFraction() float64 {
```
Comment (disksing) [inline server/cache.go]: usedFraction -> useRatio?

File: server/cache_test.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+func (s *testClusterCacheSuite) TearDownSuite(c *C) {
+	s.svr.Close()
+	s.client.Close()
+}
+
+func mustEqualStore(c *C, store1 *metapb.Store, store2 *metapb.Store) {
+	c.Assert(store1.GetId(), Equals, store2.GetId())
+	c.Assert(store1.GetAddress(), Equals, store2.GetAddress())
+}
```
Comment (disksing) [inline server/cache_test.go]: Why not use `DeepEqual`?

Comment (qiuyesuifeng) [inline server/cache_test.go]: The value type of `store1.GetAddress()` is `string`, so Equal is OK here.

File: server/cache.go (hunk 6 above)
Comment (qiuyesuifeng) [inline server/cache.go]: @queenypingcap What is your opinion?

File: server/cache_test.go (hunk 7 above)
Comment (disksing) [inline server/cache_test.go]: I mean you could use `c.Assert(store1, DeepEquals, store2)`, don't have to use these helper functions.

Comment (qiuyesuifeng) [inline server/cache_test.go]: Gotcha, a good suggestion.

**Comment by disksing** [comment]
LGTM

File: server/cache.go (hunk 6 above)
Comment (queenypingcap) [inline server/cache.go]: Yes, please use Ratio.

File: server/cache.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+			if storeExist {
+				delete(storeRegions, regionID)
+				if len(storeRegions) == 0 {
+					delete(r.storeLeaderRegions, storeID)
+				}
+			}
+		}
+	}
+
+	r.leaderRegions[regionID] = &RegionInfo{
```
Comment (siddontang) [inline server/cache.go]: If version or conf version doesn't change, I think we can return directly, no need to update cache again.

Comment (qiuyesuifeng) [inline server/cache.go]: Agreed, we should add this check later, maybe in next PR.

Comment (siddontang) [inline server/cache.go]: so please add a TODO comment here.

**Comment by siddontang** [comment]
Rest LGTM

**Comment by qiuyesuifeng** [comment]
@siddontang PTAL

**Pull request merged**